nixos/mysql: don't run parts of mysqld.service as root

flokli opened the pull request:
Also don't use `RuntimeDirectory=mysqld`, which translates to
`/run/mysqld`, which is `cfg.pidDir`, but only if `cfg.pidDir` isn't set
to something else.

Use `systemd.tmpfiles.rules` to create `dataDir` and `pidDir`, instead
of relying on the deprecated PermissionsStartOnly=yes [0][1] to setup
filesystem structure.

Also group some of the different steps together for better readability.

###### Motivation for this change
[0] https://github.com/systemd/systemd/pull/10802
[1] https://github.com/NixOS/nixpkgs/issues/53852

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

flokli commented on nixos/modules/services/databases/mysql.nix:
```diff
@@ -313,68 +318,71 @@ in
+            cmdWatchForMysqlSocket = ''
+              # Wait until the MySQL server is available for use
+              count=0
+              while [ ! -e /run/mysqld/mysqld.sock ]
```
As stated above, I'm not sure whether this is always `/run/mysqld` - it doesn't look like this depends on `pidDir`.

flokli commented:
Ran the `mysql` tests successfully, `mysqlReplication` is broken since December 2018 and reported here: https://github.com/NixOS/nixpkgs/issues/61523

flokli committed:
nixos/mysql: use systemd.tmpfiles to setup dataDir and pidDir

We need to keep using `RuntimeDirectory=mysqld`, which translates to
`/run/mysqld`, as this is used for the location of the file socket, that
could differ with what is configured via `cfg.pidDir`.

flokli committed:
nixos/mysql: reformat, move logical steps into variables

define commands like "waiting for the mysql socket to appear" or "setup
initial databases" in a let expression, so the main control flow becomes
more readable.

flokli committed:
nixos/mysql: run as mysql user and group

As we don't need to setup data directories from ExecStartPre= scripts
anymore, which required root, but use systemd.tmpfiles.rules instead,
everything can be run as just the mysql user.

flokli commented:
Splitted the changes into multiple commits as suggested by @aanderse over IRC.

aanderse commented on nixos/modules/services/databases/mysql.nix:
```diff
@@ -313,68 +318,71 @@ in
-            chown -R ${cfg.user} ${cfg.pidDir}
-          '';
+        preStart = ''
+        if ! test -e ${cfg.dataDir}/mysql; then
```
missed an indent

flokli replied:
thanks! fixed.

aanderse reviewed (approved):
Aside from running as this as a different user the changes are minimal and for the most part compatible.
LGTM :+1:

flokli committed:
nixos/mysql: drop services.mysql.pidDir

mysql already has its socket path hardcoded to to
/run/mysqld/mysqld.sock.
There's not much value in making the pidDir configurable, which also
points to /run/mysqld by default.

We only seem to use `services.mysql.pidDir` in the wordpress startup
script, to wait for mysql to boot up, but we can also simply wait on the
(hardcoded) socket location too.

A much nicer way to accomplish that would be to properly describe a
dependency on mysqld.service. This however is not easily doable, due to
how the apache-httpd module was designed.

flokli commented:
I did some deeper digging into the remaining `pidDir` TODO, and noticed this was only used in the wordpress module (to wait for mysql to boot up).

As we can simply wait for the mysql socket file to appear too, I dropped pidDir, further simplifying the mysql module.

Also added release not entries for both changes.

Izorkin commented:
Checked on mariadb-galera server- worked. Thanks.

flokli commented:
Rebased on latest master, tests still pass.

flokli commented:
@thorstenweber83 we have a nixos test using `services.mysql.ensure(Users,Databases)`, which didn't break after this PR, so I guess this should work :-)

aanderse commented:
> Or did i miss something?

@thorstenweber83 Yes. The part you missed is that `mysql` on NixOS is insanely insecure out of the box and allows **anyone** to authenticate as the `root` `mysql` user. If you were to secure your `mysql` setup, which I hope you and everyone else has, this change will then likely (depending on how you secured) cause breakage for `ensureUsers` and `ensureDatabases` as you suggested.

I have been meaning to address the overall issue of an insecure `mysql` setup out of the box on NixOS for a while. I'll discuss with @flokli and will report back.

thorstenweber83 commented:
@aanderse you're right, I use a postStart script to secure access like mysql_secure_installation does and set up socket authentication for root:

```nix
systemd.services.mysql.postStart = ''
    ( echo "ALTER USER root@localhost IDENTIFIED WITH unix_socket;"
      echo "ALTER USER root@'127.0.0.1' IDENTIFIED WITH unix_socket;"
      echo "ALTER USER root@'::1' IDENTIFIED WITH unix_socket;"

      echo "DELETE FROM mysql.user WHERE User=''';"
      echo "DELETE FROM mysql.user WHERE User='root' AND Host NOT IN ('localhost', '127.0.0.1', '::1');"
      echo "DROP DATABASE IF EXISTS test;"
      echo "DELETE FROM mysql.db WHERE Db='test' OR Db='test\\_%';"
      echo "FLUSH PRIVILEGES;"
    ) | ${mysql}/bin/mysql -u root -N
  '';
```

aanderse commented:
>       echo "ALTER USER root@'127.0.0.1' IDENTIFIED WITH unix_socket;"

I wouldn't have thought that to make sense as `mysql`/`mariadb` can only do socket authentication on `localhost`.

thorstenweber83 commented:
@aanderse yeah you're right, this is a bit nonsensical.

aanderse commented:
Partial revert. I don't have the time/energy to go any further on this at the moment. Important to get this fixed now, though.

See https://github.com/NixOS/nixpkgs/commit/4243732992b6a0cbafefe17c57c91adda86e86b5 for proposal. Note that this is based on https://github.com/NixOS/nixpkgs/pull/63786 which I'll wait to be merged before going ahead with this.

ping @flokli and @thorstenweber83
